particle-system extension

sushantojal opened the pull request:
An extension with a very basic particle system. Renders the vertices of GVRMesh as particles. Currently only  two emitter shapes ( plane and sphere ) are supported.  Both continious and bursty emission of particles are supported. The apis exposed can be seen in the GVRf demo: gvr-particles.  The current limit is approximately around 10k particles with a per second particle emission of around  400-500.

NolaDonato commented:
It would be nice to support a texture atlas to allow particles from the same emitter to have different textures as well as different colors.

roshanch commented on GVRf/Extensions/gvrf-particlesystem/src/main/res/raw/particle_vert.vsh:
```diff
@@ -0,0 +1,33 @@
+
+    gl_PointSize = (u_particle_size + u_size_change_rate * deltaTime);
+
+    if ( gl_PointSize < 1.0f )
```
Check if you can use clamp() instead of branching.

roshanch commented on GVRf/Extensions/gvrf-particlesystem/src/main/res/raw/particle_frag.fsh:
```diff
@@ -0,0 +1,26 @@
+
+    if ( deltaTime < 0.0f || deltaTime > u_particle_age) {
+        discard;
+    }
```
try not to use "discard" in fragment shader, it may disable early depth test.

roshanch replied:
use gl_Position = vec4(2.0,2.0,2.0,1.0) to clip the vertex instead of fragment.

liaxim commented on GVRf/Extensions/gvrf-particlesystem/src/main/java/org/gearvrf/particlesystem/Particles.java:
```diff
@@ -0,0 +1,124 @@
+ *  the actual particles when rendered using GL_POINTS.
+ */
+
+public class Particles extends GVRMesh {
```
Does this class have to derive from GVRMesh given the makeParticleMesh method? Isn't Particles more of a factory for meshes than a mesh?

sushantojal replied:
Yes true,  should have removed that. An earlier version was using this class as a GVRMesh.

liaxim replied:
Still derives from GVRMesh. Also it looks like it should be package private class in the emitters package. Are end users supposed to create meshes themselves directly?

sushantojal replied:
modified in the latest commit

liaxim commented on GVRf/Extensions/gvrf-particlesystem/src/main/java/org/gearvrf/particlesystem/emitters/GVREmitter.java:
```diff
@@ -0,0 +1,370 @@
+ *
+ */
+
+public class GVREmitter extends GVRSceneObject implements GVRDrawFrameListener {
```
Please implement the DrawFrameListener as an inner class to avoid exposing unnecessarily onDrawFrame as public method.

NolaDonato replied:
All the other scene objects (GVRCameraSceneObject, GVRWebViewSceneObject, GVRTextViewSceneObject, GVRVVideoSceneObject) implement it this way. They do:

gvrContext.registerDrawFrameListener(this);

and then implement the onDrawFrame function. Are you suggesting that we change the way we do onDrawFrame everywhere? I thought that registering the scene object as the draw frame listener was the right way to get GearVRF to call onDrawFrame. Because SceneObject implements DrawFrameListener, it already has a public onDrawFrame function.

liaxim replied:
GVRTextViewSceneObject for example uses an inner class. So the method is hidden from the public API and doesn't have to be documented. Also, technically, if the method is exposed then the implementer needs to worry about the user calling the method directly from any thread. GVRSceneObject does not implement the DrawFrameListener interface.

NolaDonato replied:
You would like for us to eliminate the pattern below from our code? I found two places but there are more...
public class GVRCameraSceneObject extends GVRSceneObject implement GVRDrawFrameListener
public class GVRWebViewSceneObject extends GVRSceneObject implements GVRDrawFrameListener

nitosan commented:
What kind of particle editor does this support?

sushantojal commented:
Don't have one yet, but we're considering making one! Right now, the parameters needs to be changed from the application and rebuilt.

sushantojal commented:
Changes made:
removed visibility of onDrawFrame from the application.
class 'Particles' does not extend from GVRMesh.
clipping vertices in the vertex shader instead of discarding fragments in the fragment shader.
removed unnecessary branching from shader.

liaxim commented:
Tried to run the demo app on S8+. Shader errors:

```
07-26 00:54:23.295 16902 16937 E gvrf    : SHADER: end added custom shader 1000
07-26 00:54:23.297 16902 16937 D gvrf    : SHADER: Generated lit material shader #1000 ColorShader$
07-26 00:54:23.297 16902 16937 V gvrf    : CustomShader::texture:add variable: main_texture key: main_texture
07-26 00:54:23.298 16902 16937 V gvrf    : CustomShader::uniform:add variable: u_color key: u_color
07-26 00:54:23.298 16902 16937 V gvrf    : CustomShader::uniform:add variable: u_particle_age key: u_particle_age
07-26 00:54:23.298 16902 16937 V gvrf    : CustomShader::uniform:add variable: u_acceleration key: u_acceleration
07-26 00:54:23.298 16902 16937 V gvrf    : CustomShader::uniform:add variable: u_size_change_rate key: u_size_change_rate
07-26 00:54:23.298 16902 16937 V gvrf    : CustomShader::uniform:add variable: u_time key: u_time
07-26 00:54:23.299 16902 16937 V gvrf    : CustomShader::uniform:add variable: u_particle_size key: u_particle_size
07-26 00:54:23.299 16902 16937 V gvrf    : CustomShader::uniform:add variable: u_fade key: u_fade
07-26 00:54:23.301 16902 16937 E gvrf    : Could not compile shader 35633:
07-26 00:54:23.301 16902 16937 E gvrf    : 0:22: L0002: Undeclared variable 'f'
07-26 00:54:23.301 16902 16937 E gvrf    : 0:27: L0002: Undeclared variable 'pos'
07-26 00:54:23.301 16902 16937 E gvrf    : 0:29: L0002: Undeclared variable 'f'
07-26 00:54:23.301 16902 16937 E gvrf    : 0:31: L0002: Undeclared variable 'f'
07-26 00:54:23.301 16902 16937 E gvrf    : Custom shader added program 0
07-26 00:54:23.301 16902 16937 V gvrf    : CustomShader::texture:location: variable: main_texture location: -1
07-26 00:54:23.301 16902 16937 E gvrf    : CustomShader::initialize texture error: GL_INVALID_VALUE
07-26 00
```

liaxim commented:
Good. Now please create the corresponding pull request(s) for the vulkan branch.

dnvasilev commented on GVRf/Extensions/gvrf-particlesystem/build.gradle:
```diff
@@ -0,0 +1,90 @@
+
+    sourceSets.main {
+        java.srcDirs = ['src/main/java']
+        //jni.srcDirs = [] // no auto generation of Android.mk
```
Mysterious commented-out code in gradle file. If we don't clean it up now, it will stay here forever and fossilize.

dnvasilev commented on GVRf/Extensions/gvrf-particlesystem/src/main/java/org/gearvrf/particlesystem/GVREmitter.java:
```diff
@@ -0,0 +1,350 @@
+    {
+        super(gvrContext);
+        mGVRContext = gvrContext;
+//        mGVRContext.registerDrawFrameListener(this);
```
Commented out line - why?

liaxim commented:
Running on S8+ yields this:
```
07-29 05:05:36.681 22888 22970 W System.err: java.lang.IndexOutOfBoundsException: Index: 11, Size: 11
07-29 05:05:36.682 22888 22970 W System.err: 	at java.util.ArrayList.get(ArrayList.java:411)
07-29 05:05:36.682 22888 22970 W System.err: 	at org.gearvrf.particlesystem.GVREmitter.onDrawFrame(GVREmitter.java:111)
07-29 05:05:36.682 22888 22970 W System.err: 	at org.gearvrf.particlesystem.GVRPlaneEmitter$GVRDrawFrameListenerImpl.onDrawFrame(GVRPlaneEmitter.java:153)
07-29 05:05:36.682 22888 22970 W System.err: 	at org.gearvrf.GVRViewManager.doMemoryManagementAndPerFrameCallbacks(GVRViewManager.java:259)
07-29 05:05:36.682 22888 22970 W System.err: 	at org.gearvrf.GVRViewManager.access$600(GVRViewManager.java:46)
07-29 05:05:36.682 22888 22970 W System.err: 	at org.gearvrf.GVRViewManager$3.beforeDrawEyes(GVRViewManager.java:434)
07-29 05:05:36.682 22888 22970 W System.err: 	at org.gearvrf.GVRViewManager.beforeDrawEyes(GVRViewManager.java:491)
07-29 05:05:36.682 22888 22970 W System.err: 	at org.gearvrf.OvrViewManager.beforeDrawEyes(OvrViewManager.java:205)
07-29 05:05:36.682 22888 22970 W System.err: 	at org.gearvrf.OvrViewManager.onDrawFrame(OvrViewManager.java:258)
07-29 05:05:36.682 22888 22970 W System.err: 	at org.gearvrf.OvrVrapiActivityHandler$7.onDrawFrame(OvrVrapiActivityHandler.java:437)
07-29 05:05:36.682 22888 22970 W System.err: 	at android.opengl.GLSurfaceView$GLThread.guardedRun(GLSurfaceView.java:1562)
07-29 05:05:36.682 22888 22970 W System.err: 	at android.opengl.GLSurfaceView$GLThread.run(GLSurfaceView.java:1262)
```

liaxim commented on GVRf/Extensions/gvrf-particlesystem/src/main/java/org/gearvrf/particlesystem/ColorShader.java:
```diff
@@ -0,0 +1,39 @@
+import org.gearvrf.GVRShaderTemplate;
+import org.gearvrf.utility.TextFile;
+
+public class ColorShader extends GVRShaderTemplate
```
Please remove the public modifier.

sushantojal replied:
Doing this gives the following error:
java.lang.UnsupportedOperationException: error creating shader template of class ParticleShaderjava.lang.Class<org.gearvrf.particlesystem.ParticleShader> is not accessible from java.lang.Class<org.gearvrf.GVRBaseShaderManager>
The call causing this error is "retrieveShaderTemplate" in from an instance of GVRMaterialShaderManager in Particles.java.
`        GVRMaterialShaderManager shaderManager = mGVRContext.getMaterialShaderManager();
        GVRShaderTemplate shaderTemplate = shaderManager.retrieveShaderTemplate(ParticleShader.class);
        shaderTemplate.bindShader(mGVRContext, meshObject.getRenderData(), mGVRContext.getMainScene());`
Looks like any custom shader class needs to be visible to the framework.

liaxim replied:
Makes sense.

liaxim commented on GVRf/Extensions/gvrf-particlesystem/src/main/java/org/gearvrf/particlesystem/GVREmitter.java:
```diff
@@ -0,0 +1,349 @@
+     * @param depth volume depth (along z-axis)
+     */
+
+    public void setParticleVolume(float width, float height, float depth)
```
Since this can be called from any method it is unlikely to be thread-safe. In this particular case you could just run it on the gl thread.

liaxim commented on GVRf/Extensions/gvrf-particlesystem/src/main/java/org/gearvrf/particlesystem/GVREmitter.java:
```diff
@@ -0,0 +1,349 @@
+     * @param minV Minimum velocity that a particle can have
+     * @param maxV Maximum velocity that a particle can have
+     */
+    public void setVelocityRange( float minV, float maxV )
```
There is no getter so you could make it more thread-safe by running the body on the gl thread.

sushantojal replied:
Should this be done for all the public APIs which set the properties?

liaxim replied:
We'll get back to thread-safety in the future.

liaxim commented on GVRf/Extensions/gvrf-particlesystem/src/main/java/org/gearvrf/particlesystem/GVREmitter.java:
```diff
@@ -0,0 +1,349 @@
+
+    //list of active scene objects containing particle meshes,
+    // and their corresponding spawn times.
+    protected ArrayList<GVRSceneObject> mActiveParticleMeshes = null;
```
Seems like these two contain pairs of associated object and time. You could use one container and android.util.Pair to make the link more explicit.

liaxim commented on GVRf/Extensions/gvrf-particlesystem/src/main/java/org/gearvrf/particlesystem/GVREmitter.java:
```diff
@@ -0,0 +1,349 @@
+        mColor = color;
+    }
+
+    public void clearSystem()
```
Public methods need javadoc. Since this is the base class it really needs one. The derived would automatically inherit the javadoc supposedly. Also there is the {@inheritDoc} tag.

liaxim commented on GVRf/Extensions/gvrf-particlesystem/src/main/java/org/gearvrf/particlesystem/GVRPlaneEmitter.java:
```diff
@@ -0,0 +1,191 @@
+
+import java.lang.ref.WeakReference;
+import java.util.Random;
+
```
I think the javadoc should go right in front of the class to be picked up by the generator.

sushantojal commented:
Made the requested modifications in the latest commit. Still working to make it work with the vulkan branch. Once that is done, will create a PR for vulkan branch with these changes included.

liaxim commented:
Ok. Works now. There are occasional exceptions like this one but you could look into them later.

```
08-03 07:49:54.517 9231-9403/org.gearvrf.particles D/gvrf: SHADER: Reuse shader #1000 ParticleShader$
08-03 07:49:54.517 9231-9403/org.gearvrf.particles E/GVRViewManager: DrawFrameListener org.gearvrf.particlesystem.GVRPlaneEmitter$GVRDrawFrameListenerImpl@5fb2a27 threw java.lang.IndexOutOfBoundsException: Index: 4, Size: 4
08-03 07:49:54.517 9231-9403/org.gearvrf.particles W/System.err: java.lang.IndexOutOfBoundsException: Index: 4, Size: 4
08-03 07:49:54.517 9231-9403/org.gearvrf.particles W/System.err:     at java.util.ArrayList.get(ArrayList.java:411)
08-03 07:49:54.517 9231-9403/org.gearvrf.particles W/System.err:     at org.gearvrf.particlesystem.GVREmitter.onDrawFrame(GVREmitter.java:108)
08-03 07:49:54.518 9231-9403/org.gearvrf.particles W/System.err:     at org.gearvrf.particlesystem.GVRPlaneEmitter$GVRDrawFrameListenerImpl.onDrawFrame(GVRPlaneEmitter.java:153)
08-03 07:49:54.518 9231-9403/org.gearvrf.particles W/System.err:     at org.gearvrf.GVRViewManager.doMemoryManagementAndPerFrameCallbacks(GVRViewManager.java:259)
08-03 07:49:54.518 9231-9403/org.gearvrf.particles W/System.err:     at org.gearvrf.GVRViewManager.access$600(GVRViewManager.java:46)
08-03 07:49:54.518 9231-9403/org.gearvrf.particles W/System.err:     at org.gearvrf.GVRViewManager$3.beforeDrawEyes(GVRViewManager.java:434)
08-03 07:49:54.518 9231-9403/org.gearvrf.particles W/System.err:     at org.gearvrf.GVRViewManager.beforeDrawEyes(GVRViewManager.java:491)
08-03 07:49:54.518 9231-9403/org.gearvrf.particles W/System.err:     at org.gearvrf.OvrViewManager.beforeDrawEyes(OvrViewManager.java:205)
08-03 07:49:54.518 9231-9403/org.gearvrf.particles W/System.err:     at org.gearvrf.OvrViewManager.onDrawFrame(OvrViewManager.java:258)
08-03 07:49:54.518 9231-9403/org.gearvrf.particles W/System.err:     at org.gearvrf.OvrVrapiActivityHandler$7.onDrawFrame(OvrVrapiActivityHandler.java:437)
08-03 07:49:54.518 9231-9403/org.gearvrf.particles W/System.err:     at android.opengl.GLSurfaceView$GLThread.guardedRun(GLSurfaceView.java:1562)
08-03 07:49:54.518 9231-9403/org.gearvrf.particles W/System.err:     at android.opengl.GLSurfaceView$GLThread.run(GLSurfaceView.java:1262)
```

liaxim commented on GVRf/Extensions/gvrf-particlesystem/src/main/java/org/gearvrf/particlesystem/GVREmitter.java:
```diff
@@ -0,0 +1,365 @@
+            if (currTime - currMeshSpawnTime > mMaxAge + 1) {
+                GVRSceneObject toDelete = meshInfo.get(i).first;
+                GVREmitter.this.removeChildObject(toDelete);
+                meshInfo.remove(i);
```
You are removing something from the collection so the size changes. But you still iterate over the old size of the collection. This is the source of the exceptions I mentioned. Seem to be non-critical but ideally you want it fixed.

sushantojal replied:
Yes, makes sense. Will fix it right away.

liaxim commented:
Looks ok to me. Planning to merge on Monday.

liaxim commented:
Needs DCO

sushantojal commented:
GearVRf-DCO-1.0-Signed-off-by: Sushant Ojal
sushant.o@partner.samsung.com
